src/struct_mv/struct_matvec_core.c

```diff
@@ -0,0 +1,634 @@
+ * Note: This function computes -A*x.
+ *--------------------------------------------------------------------------*/
+
+HYPRE_Int
+hypre_StructMatvecCompute_core_CC( hypre_StructMatrix *A,
+                                   hypre_StructVector *x,
+                                   hypre_StructVector *y,
+                                   HYPRE_Int           Ab,
+                                   HYPRE_Int           xb,
+                                   HYPRE_Int           yb,
+                                   HYPRE_Int           transpose,
+                                   HYPRE_Int           nentries,
+                                   HYPRE_Int          *entries,
+                                   hypre_IndexRef      start,
+                                   hypre_IndexRef      stride,
+                                   hypre_IndexRef      loop_size,
+                                   hypre_IndexRef      xfstride,
+                                   hypre_IndexRef      ran_stride,
+                                   hypre_IndexRef      xdstride,
+                                   hypre_IndexRef      ydstride,
+                                   hypre_Box          *x_data_box,
+                                   hypre_Box          *y_data_box )
+{
+   HYPRE_Int             ndim = hypre_StructMatrixNDim(A);
+   hypre_StructStencil  *stencil = hypre_StructMatrixStencil(A);
+   hypre_Index          *stencil_shape = hypre_StructStencilShape(stencil);
+
+   hypre_Index           xdstart, ydstart;
+   hypre_Index           offset;
+
+   HYPRE_Complex        *Ap0, *Ap1, *Ap2;
+   HYPRE_Complex        *Ap3, *Ap4, *Ap5;
+   HYPRE_Complex        *Ap6, *Ap7, *Ap8;
+   HYPRE_Complex        *xp,  *yp;
+   HYPRE_Int             xoff0, xoff1, xoff2;
+   HYPRE_Int             xoff3, xoff4, xoff5;
+   HYPRE_Int             xoff6, xoff7, xoff8;
+
+   HYPRE_Int             si, depth;
+
+   HYPRE_ANNOTATE_FUNC_BEGIN;
+
+   xp = hypre_StructVectorBoxData(x, xb);
+   yp = hypre_StructVectorBoxData(y, yb);
+
+   hypre_CopyToIndex(start, ndim, xdstart);
+   /* The next line is only used to avoid 'print error' messages.  It ensures
+    * that xdstart is aligned with the vector data space before mapping.  The
+    * choice, Neg vs Pos, doesn't matter because an offset will be used to index
+    * into the vector x (xoff = index - xdstart). */
+   hypre_SnapIndexNeg(xdstart, NULL, stride, ndim);
+   hypre_MapToFineIndex(xdstart, NULL, xfstride, ndim);
+   hypre_StructVectorMapDataIndex(x, xdstart);
+   hypre_CopyToIndex(start, ndim, ydstart);
+   hypre_MapToCoarseIndex(ydstart, NULL, ran_stride, ndim);
+
+#define DEVICE_VAR is_device_ptr(yp,xp)
+   for (si = 0; si < nentries; si += UNROLL_MAXDEPTH)
+   {
+      depth = hypre_min(UNROLL_MAXDEPTH, (nentries - si));
+
+      if (!transpose)
+      {
+         switch (depth)
+         {
+            case 9:
+               HYPRE_SET_CAX(Ap8, xoff8, entries[si + 8]);
```

Comment:
```suggestion
               HYPRE_SET_CAX(Ap8, xoff8, entries[si + 8]);
               HYPRE_FALLTHROUGH;
```

Should we do the same for the other occurrences in this file as well?

Reply:
I can't implement FALLTHROUGH until I pull in the recmat changes, but I wanted to wait until the GPU work was all integrated (or mostly integrated).  If you think it's okay to do the merge now, let me know.

Reply:
This is done now